### PR TITLE
Rewrite mentions input using typeahead.js

### LINKED ISSUE #4786: wrong placement of @-mentions in a post
When writing a post in which I mention a user whose username start with the same letter/s as for example a diaspora id then the @-mention will replace the corresponding @ and the following letters in the diaspora id. This results in a totally wrong placement. A workaround is to make the @-mention first and add the diaspora id after.

I faced this issue when creating the following post: https://diasp.eu/posts/34d40ce26c5c4a13

Here the @ and "dias" of the ID "anewnamnevi @ diasp.org" collided with the "@DiasporaHQ" mention as seen below:

wrong mention pacement

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1413411-wrong-placement-of-mentions-in-a-post?utm_campaign=plugin&utm_content=tracker%2F522&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F522&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by svbergerem**
Supersedes #6537, fixes #4786.

I based this PR on #6537 and refactored the code. I hope that this is now much more readable and maintainable.

### CODE REVIEW

**Comment by AugierLe42e** [comment]
Nice! Thanks for taking care of this, @svbergerem ;)

File: app/assets/javascripts/app/views/aspect_create_view.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ app.views.AspectCreate = app.views.Base.extend({
   },
 
   inputKeypress: function(evt) {
-    if(evt.which === 13) {
+    if(evt.which === Keycodes.ENTER) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/aspect_create_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/comment_stream_view.js [hunk 2]
```diff
@@ -56,7 +56,7 @@ app.views.CommentStream = app.views.Base.extend({
   },
 
   keyDownOnCommentBox: function(evt) {
-    if(evt.keyCode === 13 && evt.ctrlKey) {
+    if(evt.which === Keycodes.ENTER && evt.ctrlKey) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/comment_stream_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 3]
```diff
@@ -43,7 +43,7 @@ app.views.ConversationsForm = Backbone.View.extend({
   },
 
   keyDown : function(evt) {
-    if( evt.keyCode === 13 && evt.ctrlKey ) {
+    if(evt.which === Keycodes.ENTER && evt.ctrlKey) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_form_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/conversations_view.js [hunk 4]
```diff
@@ -50,7 +50,7 @@ app.views.Conversations = Backbone.View.extend({
   },
 
   keyDown : function(evt) {
-    if( evt.keyCode === 13 && evt.ctrlKey ) {
+    if(evt.which === Keycodes.ENTER && evt.ctrlKey) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      self.updateMessageTexts();
+    });
+  },
+
+  /**
+   * Selects next or previous result when result dropdown is open and
+   * user press up and down arrows.
+   */
+  onArrowKeyDown: function(e){
+    if(!this.isVisible() || (e.which !== Keycodes.UP && e.which !== Keycodes.DOWN)){
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+   */
+  onInputBoxInput: function() {
+    this.cleanMentionedPeople();
+    this.updateMessageTexts();
+    this.updateTypeaheadInput();
+  },
+
+  onInputBoxKeyDown: function(e){
+    // This also matches HOME/END on OSX which is CMD+LEFT, CMD+RIGHT
+    if(e.which === Keycodes.LEFT || e.which === Keycodes.RIGHT ||
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+       e.which === Keycodes.HOME || e.which === Keycodes.END){
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+      _.defer(_.bind(this.updateTypeaheadInput, this));
+      return;
+    }
+
+    if(!this.isVisible){
+      return true;
+    }
+
+    switch(e.which){
+      case Keycodes.ESC:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+      case Keycodes.SPACE:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        this.closeSuggestions();
+        break;
+      case Keycodes.UP:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+      case Keycodes.DOWN:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        this.onArrowKeyDown(e);
+        break;
+      case Keycodes.RETURN:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+      case Keycodes.TAB:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher_view.js [hunk 14]
```diff
@@ -244,7 +249,7 @@ app.views.Publisher = Backbone.View.extend({
 
   // avoid submitting form when pressing Enter key
   avoidEnter: function(evt){
-    if(evt.keyCode === 13)
+    if(evt.which === Keycodes.ENTER)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher_view.js [hunk 15]
```diff
@@ -379,7 +358,7 @@ app.views.Publisher = Backbone.View.extend({
   },
 
   keyDown : function(evt) {
-    if( evt.keyCode === 13 && evt.ctrlKey ) {
+    if(evt.which === Keycodes.ENTER && evt.ctrlKey) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/search_view.js [hunk 16]
```diff
@@ -1,96 +1,35 @@
...
   },
 
-  suggestionSelected: function(evt, datum) {
-    window.location = datum.url;
-  },
-
-  inputKeypress: function(evt) {
-    if(evt.which === 13 && $(".tt-suggestion.tt-cursor").length === 0) {
+  inputKeypress: function(evt){
+    if(evt.which === Keycodes.ENTER && $(".tt-suggestion.tt-cursor").length === 0){
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/search_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/tag_following_list_view.js [hunk 17]
```diff
@@ -46,7 +46,7 @@ app.views.TagFollowingList = app.views.Base.extend({
     });
 
     this.$("input").bind('keydown', function(evt){
-      if(evt.keyCode === 13 || evt.keyCode === 9 || evt.keyCode === 32){
+      if(evt.which === Keycodes.ENTER || evt.which === Keycodes.TAB || evt.which === Keycodes.SPACE) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/tag_following_list_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/pages/users-getting-started.js [hunk 18]
```diff
@@ -73,7 +73,7 @@ Diaspora.Pages.UsersGettingStarted = function() {
       });
 
     autocompleteInput.bind('keydown', function(evt){
-      if(evt.keyCode === 13 || evt.keyCode === 9 || evt.keyCode === 32){
+      if(evt.which === Keycodes.ENTER || evt.which === Keycodes.TAB || evt.which === Keycodes.SPACE) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/pages/users-getting-started.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/aspect_create_view_spec.js [hunk 19]
```diff
@@ -47,13 +47,13 @@ describe("app.views.AspectCreate", function() {
       });
 
       it("should call createAspect if the enter key was pressed", function() {
-        var e = $.Event("keypress", { which: 13 });
+        var e = $.Event("keypress", { which: Keycodes.ENTER });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/aspect_create_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/aspect_create_view_spec.js [hunk 20]
```diff
@@ -47,13 +47,13 @@ describe("app.views.AspectCreate", function() {
...
         this.view.inputKeypress(e);
         expect(this.view.createAspect).toHaveBeenCalled();
       });
 
       it("shouldn't call createAspect if another key was pressed", function() {
-        var e = $.Event("keypress", { which: 42 });
+        var e = $.Event("keypress", { which: Keycodes.TAB });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/aspect_create_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/comment_stream_view_spec.js [hunk 21]
```diff
@@ -107,8 +107,7 @@ describe("app.views.CommentStream", function(){
       var form = this.view.$("form");
       form.submit(submitCallback);
 
-      var e = $.Event("keydown", { keyCode: 13 });
-      e.ctrlKey = false;
+      var e = $.Event("keydown", { which: Keycodes.ENTER, ctrlKey: false });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/comment_stream_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/comment_stream_view_spec.js [hunk 22]
```diff
@@ -119,8 +118,7 @@ describe("app.views.CommentStream", function(){
       var form = this.view.$("form");
       form.submit(submitCallback);
 
-      var e = $.Event("keydown", { keyCode: 13 });
-      e.ctrlKey = true;
+      var e = $.Event("keydown", { which: Keycodes.ENTER, ctrlKey: true });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/comment_stream_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/conversations_view_spec.js [hunk 23]
```diff
@@ -64,14 +64,14 @@ describe("app.views.Conversations", function(){
 
     it("should submit the form with ctrl+enter", function(){
       $("form#new_message").submit(this.submitCallback);
-      var e = $.Event("keydown", { keyCode: 13, ctrlKey: true });
+      var e = $.Event("keydown", { which: Keycodes.ENTER, ctrlKey: true });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/conversations_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/conversations_view_spec.js [hunk 24]
```diff
@@ -64,14 +64,14 @@ describe("app.views.Conversations", function(){
...
       $("textarea#message_text").trigger(e);
       expect(this.submitCallback).toHaveBeenCalled();
     });
 
     it("shouldn't submit the form without the ctrl key", function(){
       $("form#new_message").submit(this.submitCallback);
-      var e = $.Event("keydown", { keyCode: 13, ctrlKey: false });
+      var e = $.Event("keydown", { which: Keycodes.ENTER, ctrlKey: false });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/conversations_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/publisher_mention_view_spec.js [hunk 25]
```diff
@@ -0,0 +1,472 @@
...
+  });
+
+  describe("onInputBoxKeyDown", function() {
+    beforeEach(function() {
+      this.view = new app.views.PublisherMention({ el: "#publisher" });
+    });
+
+    context("escape key", function() {
+      beforeEach(function() {
+        this.evt = $.Event("keydown", {which: Keycodes.ESC});
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/publisher_mention_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/publisher_mention_view_spec.js [hunk 26]
```diff
@@ -0,0 +1,472 @@
...
+      it("calls 'closeSuggestions'", function() {
+        spyOn(this.view, "closeSuggestions");
+        this.view.onInputBoxKeyDown(this.evt);
+        expect(this.view.closeSuggestions).toHaveBeenCalled();
+      });
+    });
+
+    context("space key", function() {
+      beforeEach(function() {
+        this.evt = $.Event("keydown", {which: Keycodes.SPACE});
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/publisher_mention_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/publisher_mention_view_spec.js [hunk 27]
```diff
@@ -0,0 +1,472 @@
...
+      it("calls 'closeSuggestions'", function() {
+        spyOn(this.view, "closeSuggestions");
+        this.view.onInputBoxKeyDown(this.evt);
+        expect(this.view.closeSuggestions).toHaveBeenCalled();
+      });
+    });
+
+    context("up key", function() {
+      beforeEach(function() {
+        this.evt = $.Event("keydown", {which: Keycodes.UP});
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/publisher_mention_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/publisher_mention_view_spec.js [hunk 28]
```diff
@@ -0,0 +1,472 @@
...
+      it("calls 'onArrowKeyDown'", function() {
+        spyOn(this.view, "onArrowKeyDown");
+        this.view.onInputBoxKeyDown(this.evt);
+        expect(this.view.onArrowKeyDown).toHaveBeenCalled();
+      });
+    });
+
+    context("down key", function() {
+      beforeEach(function() {
+        this.evt = $.Event("keydown", {which: Keycodes.DOWN});
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/publisher_mention_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/publisher_mention_view_spec.js [hunk 29]
```diff
@@ -0,0 +1,472 @@
...
+      it("calls 'onArrowKeyDown'", function() {
+        spyOn(this.view, "onArrowKeyDown");
+        this.view.onInputBoxKeyDown(this.evt);
+        expect(this.view.onArrowKeyDown).toHaveBeenCalled();
+      });
+    });
+
+    context("return key", function() {
+      beforeEach(function() {
+        this.evt = $.Event("keydown", {which: Keycodes.RETURN});
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/publisher_mention_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/publisher_mention_view_spec.js [hunk 30]
```diff
@@ -0,0 +1,472 @@
...
+      it("calls 'onSuggestionSelection'", function() {
+        spyOn(this.view, "onSuggestionSelection");
+        this.view.onInputBoxKeyDown(this.evt);
+        expect(this.view.onSuggestionSelection).toHaveBeenCalled();
+      });
+    });
+
+    context("tab key", function() {
+      beforeEach(function() {
+        this.evt = $.Event("keydown", {which: Keycodes.TAB});
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/publisher_mention_view_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 31]
```diff
@@ -16,9 +16,7 @@ describe("app.views.StreamShortcuts", function () {
   describe("pressing 'j'", function(){
     it("should call 'gotoNext' if not pressed in an input field", function(){
       spyOn(this.view, 'gotoNext');
-      var e = $.Event("keydown", { which: 74, target: {type: "div"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('j');
+      var e = $.Event("keydown", { which: Keycodes.J, target: {type: "div"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 32]
```diff
@@ -32,9 +30,7 @@ describe("app.views.StreamShortcuts", function () {
     it("shouldn't do anything if the user types in an input field", function(){
       spyOn(this.view, 'gotoNext');
       spyOn(this.view, 'selectPost');
-      var e = $.Event("keydown", { which: 74, target: {type: "textarea"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('j');
+      var e = $.Event("keydown", { which: Keycodes.J, target: {type: "textarea"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 33]
```diff
@@ -44,9 +40,7 @@ describe("app.views.StreamShortcuts", function () {
   describe("pressing 'k'", function(){
     it("should call 'gotoPrev' if not pressed in an input field", function(){
       spyOn(this.view, 'gotoPrev');
-      var e = $.Event("keydown", { which: 75, target: {type: "div"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('k');
+      var e = $.Event("keydown", { which: Keycodes.K, target: {type: "div"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 34]
```diff
@@ -60,9 +54,7 @@ describe("app.views.StreamShortcuts", function () {
     it("shouldn't do anything if the user types in an input field", function(){
       spyOn(this.view, 'gotoPrev');
       spyOn(this.view, 'selectPost');
-      var e = $.Event("keydown", { which: 75, target: {type: "textarea"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('k');
+      var e = $.Event("keydown", { which: Keycodes.K, target: {type: "textarea"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 35]
```diff
@@ -72,18 +64,14 @@ describe("app.views.StreamShortcuts", function () {
   describe("pressing 'c'", function(){
     it("should click on the comment-button if not pressed in an input field", function(){
       spyOn(this.view, 'commentSelected');
-      var e = $.Event("keyup", { which: 67, target: {type: "div"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('c');
+      var e = $.Event("keyup", { which: Keycodes.C, target: {type: "div"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 36]
```diff
@@ -72,18 +64,14 @@ describe("app.views.StreamShortcuts", function () {
...
       this.view._onHotkeyUp(e);
       expect(this.view.commentSelected).toHaveBeenCalled();
     });
 
     it("shouldn't do anything if the user types in an input field", function(){
       spyOn(this.view, 'commentSelected');
-      var e = $.Event("keyup", { which: 67, target: {type: "textarea"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('c');
+      var e = $.Event("keyup", { which: Keycodes.C, target: {type: "textarea"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 37]
```diff
@@ -92,18 +80,14 @@ describe("app.views.StreamShortcuts", function () {
   describe("pressing 'l'", function(){
     it("should click on the like-button if not pressed in an input field", function(){
       spyOn(this.view, 'likeSelected');
-      var e = $.Event("keyup", { which: 76, target: {type: "div"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('l');
+      var e = $.Event("keyup", { which: Keycodes.L, target: {type: "div"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 38]
```diff
@@ -92,18 +80,14 @@ describe("app.views.StreamShortcuts", function () {
...
       this.view._onHotkeyUp(e);
       expect(this.view.likeSelected).toHaveBeenCalled();
     });
 
     it("shouldn't do anything if the user types in an input field", function(){
       spyOn(this.view, 'likeSelected');
-      var e = $.Event("keyup", { which: 76, target: {type: "textarea"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('l');
+      var e = $.Event("keyup", { which: Keycodes.L, target: {type: "textarea"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 39]
```diff
@@ -112,18 +96,14 @@ describe("app.views.StreamShortcuts", function () {
   describe("pressing 'r'", function(){
     it("should click on the reshare-button if not pressed in an input field", function(){
       spyOn(this.view, 'reshareSelected');
-      var e = $.Event("keyup", { which: 82, target: {type: "div"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('r');
+      var e = $.Event("keyup", { which: Keycodes.R, target: {type: "div"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 40]
```diff
@@ -112,18 +96,14 @@ describe("app.views.StreamShortcuts", function () {
...
       this.view._onHotkeyUp(e);
       expect(this.view.reshareSelected).toHaveBeenCalled();
     });
 
     it("shouldn't do anything if the user types in an input field", function(){
       spyOn(this.view, 'reshareSelected');
-      var e = $.Event("keyup", { which: 82, target: {type: "textarea"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('r');
+      var e = $.Event("keyup", { which: Keycodes.R, target: {type: "textarea"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 41]
```diff
@@ -132,18 +112,14 @@ describe("app.views.StreamShortcuts", function () {
   describe("pressing 'm'", function(){
     it("should click on the more-button if not pressed in an input field", function(){
       spyOn(this.view, 'expandSelected');
-      var e = $.Event("keyup", { which: 77, target: {type: "div"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('m');
+      var e = $.Event("keyup", { which: Keycodes.M, target: {type: "div"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 42]
```diff
@@ -132,18 +112,14 @@ describe("app.views.StreamShortcuts", function () {
...
       this.view._onHotkeyUp(e);
       expect(this.view.expandSelected).toHaveBeenCalled();
     });
 
     it("shouldn't do anything if the user types in an input field", function(){
       spyOn(this.view, 'expandSelected');
-      var e = $.Event("keyup", { which: 77, target: {type: "textarea"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('m');
+      var e = $.Event("keyup", { which: Keycodes.M, target: {type: "textarea"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 43]
```diff
@@ -152,18 +128,14 @@ describe("app.views.StreamShortcuts", function () {
   describe("pressing 'o'", function(){
     it("should click on the more-button if not pressed in an input field", function(){
       spyOn(this.view, 'openFirstLinkSelected');
-      var e = $.Event("keyup", { which: 79, target: {type: "div"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('o');
+      var e = $.Event("keyup", { which: Keycodes.O, target: {type: "div"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/app/views/stream/shortcuts_spec.js [hunk 44]
```diff
@@ -152,18 +128,14 @@ describe("app.views.StreamShortcuts", function () {
...
       this.view._onHotkeyUp(e);
       expect(this.view.openFirstLinkSelected).toHaveBeenCalled();
     });
 
     it("shouldn't do anything if the user types in an input field", function(){
       spyOn(this.view, 'openFirstLinkSelected');
-      var e = $.Event("keyup", { which: 79, target: {type: "textarea"} });
-      //verify that the test is correct
-      expect(String.fromCharCode( e.which ).toLowerCase()).toBe('o');
+      var e = $.Event("keyup", { which: Keycodes.O, target: {type: "textarea"} });
```
Comment (diaspora-code-review) [inline spec/javascripts/app/views/stream/shortcuts_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/lib/keycodes_spec.js [hunk 45]
```diff
@@ -0,0 +1,13 @@
+describe("Keycodes", function() {
+  it("sets the correct keycode for letters", function() {
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("").forEach(function(c) {
+      expect(String.fromCharCode(Keycodes[c])).toBe(c);
```
Comment (diaspora-code-review) [inline spec/javascripts/lib/keycodes_spec.js]: 'Keycodes' is not defined.

File: spec/javascripts/lib/keycodes_spec.js [hunk 46]
```diff
@@ -0,0 +1,13 @@
...
+    });
+  });
+
+  it("sets the correct keycode for digits", function() {
+    "0123456789".split("").forEach(function(c) {
+      expect(String.fromCharCode(Keycodes[c])).toBe(c);
```
Comment (diaspora-code-review) [inline spec/javascripts/lib/keycodes_spec.js]: 'Keycodes' is not defined.

**Comment by AugierLe42e** [comment]
Wow, @diaspora-code-review, what happened!? xD

**Comment by AugierLe42e** [comment]
@svbergerem: Something bothers me here: how to add mentions to mobile view next? Should the current code be extracted out of Backbone or duplicated in another file under `/mobile`?

**Comment by svbergerem** [comment]
> NOTE: typeahead.js is not tested on mobile browsers. [(source)](https://github.com/twitter/typeahead.js/blob/5f3787a/README.md)

There have also been some ideas on IRC to refactor the mentions syntax which would make it much easier to implement it for the mobile view. But this is not in scope of this PR. If we decide to change the syntax later then the code I wrote should be pretty easy to adapt to the new syntax.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 47]
```diff
@@ -0,0 +1,224 @@
...
+   * "Hello @{user1 ; user1@pod.tld}" in the hidden element and
+   * "Hello <strong><span>user1</span></strong>" in the element visible to the user.
+   */
+  updateMessageTexts: function(){
+    var fakeMessageText = this.inputBox.val(),
+        mentionBoxText = fakeMessageText,
+        messageText = fakeMessageText;
+
+    var self = this;
+    _.each(this.mentionedPeople, function(person){
```
Comment (denschub) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: I'm wondering if we should start using `Array.prototype.forEach()` which is supported everywhere but IE 8 and older. Reducing the overall underscore usage where it is not necessary could be a good thing. Opinions?

File: app/assets/javascripts/app/views/search_base_view.js [hunk 48]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    this.bloodhound = new Bloodhound(bloodhoundOptions);
+  },
+
+  setupCustomSearch: function() {
+    var self = this;
+    this.bloodhound.customSearch = function(query, sync, async){
+      var _sync = function(datums){
+        var results = _.reject(datums, function(datum) {
```
Comment (denschub) [inline app/assets/javascripts/app/views/search_base_view.js]: General note: Sometimes, there is a space between the closing parenthesis and the opening curly brace. Please unify. :)

File: spec/javascripts/app/views/publisher_mention_view_spec.js [hunk 49]
```diff
@@ -0,0 +1,472 @@
...
+      expect(this.view.inputBox.val()).toBe("");
+      expect(this.view.onInputBoxInput).toHaveBeenCalled();
+    });
+  });
+
+  describe("closeSuggestions", function() {
+    beforeEach(function() {
+      this.view = new app.views.PublisherMention({ el: "#publisher" });
+      this.view.bloodhound.add([
+        {"person": true, "name":"user1", "handle":"user1@pod.tld"}
```
Comment (denschub) [inline spec/javascripts/app/views/publisher_mention_view_spec.js]: Missing spaces between `:` and `"`.

**Comment by denschub** [comment]
There are some style inconsistencies but other than that, it looks really nice! I'd love to get rid of `_.each` where possible.

Also, sorry everyone for the incredible noise made by @diaspora-code-review. While @svbergerem correctly changed the linter config, but the bot is configured to always use the latest `develop` configs to avoid issues. Maybe we should consider changing that...

File: app/assets/javascripts/app/views/publisher/mention_view.js (hunk 47 above)
Comment (Flaburgan) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: +1, we do not even support IE8.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 50]
```diff
@@ -0,0 +1,222 @@
...
+      this.updateMessageTexts();
+    }, this);
+  },
+
+  /**
+   * Selects next or previous result when result dropdown is open and
+   * user press up and down arrows.
+   */
+  onArrowKeyDown: function(e) {
+    if(!this.isVisible() || (e.which !== Keycodes.UP && e.which !== Keycodes.DOWN)) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 51]
```diff
@@ -0,0 +1,222 @@
...
+   */
+  onInputBoxInput: function() {
+    this.cleanMentionedPeople();
+    this.updateMessageTexts();
+    this.updateTypeaheadInput();
+  },
+
+  onInputBoxKeyDown: function(e) {
+    // This also matches HOME/END on OSX which is CMD+LEFT, CMD+RIGHT
+    if(e.which === Keycodes.LEFT || e.which === Keycodes.RIGHT ||
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 52]
```diff
@@ -0,0 +1,222 @@
...
+       e.which === Keycodes.HOME || e.which === Keycodes.END) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 53]
```diff
@@ -0,0 +1,222 @@
...
+      _.defer(_.bind(this.updateTypeaheadInput, this));
+      return;
+    }
+
+    if(!this.isVisible) {
+      return true;
+    }
+
+    switch(e.which) {
+      case Keycodes.ESC:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/publisher/mention_view.js [hunk 54]
```diff
@@ -0,0 +1,222 @@
...
+      case Keycodes.SPACE:
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/publisher/mention_view.js]: 'Keycodes' is not defined.

File: app/assets/javascripts/app/views/search_view.js [hunk 55]
```diff
@@ -83,14 +22,14 @@ app.views.Search = app.views.Base.extend({
     $(evt.target).toggleClass("active", isActive);
   },
 
-  suggestionSelected: function(evt, datum) {
-    window.location = datum.url;
-  },
-
   inputKeypress: function(evt) {
-    if(evt.which === 13 && $(".tt-suggestion.tt-cursor").length === 0) {
+    if(evt.which === Keycodes.ENTER && $(".tt-suggestion.tt-cursor").length === 0) {
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/search_view.js]: 'Keycodes' is not defined.

**Comment by denschub** [comment]
Alllright, merged! Thank you very much for your work and your refactorings.

**Pull request merged**